Version2 development autoassess landsurface

valeriupredoi opened the pull request:
Hey fellas, I am continuing with converting autoassess diags to ESMValTool and this is the new installment wohoo, excitement (my arse) :grin: I will be ready to ask you for PR approval on this one very soon (prob tomorrow), I still need to run a couple more tests

valeriupredoi commented:
ok fellas, bloody thing works finr, so when you have time, have a look and pls accept the PR :beer: :cow2:

valeriupredoi commented:
hey @schlunma I have realized there is still quite a bit of duplication and actually the ```prsn``` is needed only for one part of the diag and not for the other, so don't worry about it just yet, I will clean up, and break up the diags so we don't have to look for files that may not exist (I actually wrote a dummy CERES-EBAF for prsn). Cheers for looking at it, but I'll spend today doing these bits and I'll ask you guys to review/test it once I'm done :grin:

valeriupredoi commented:
@schlunma @bouweandela you guys when you get some free time, I'd appreciate it if you had a look at this, Manuel, you can safely run _landsurface_snow and _landsurface_surfrad recipes now :grin:

schlunma commented:
I tried to ran them, but they failed because I have no `python27` executable:

```
2018-08-29 17:06:39,839 [21092] INFO     __main__,55    /mnt/lustre01/pf/b/b309141/ESMValTool/esmvaltool/diag_scripts/autoassess/autoassess_source/plot_norm_ac.py --exp MPI-ESM-MR --ref MPI-ESM-LR --plot /scratch/b/b309141/work_v2/recipe_autoassess_landsurface_snow_20180829_150609/plots/plot_standard/plot_landsurf_snow/Snow_Metrics.png --title Landsurface_Snow_MPI-ESM-LR_vs_MPI-ESM-MR --file-exp /scratch/b/b309141/work_v2/recipe_autoassess_landsurface_snow_20180829_150609/plots/aa_landsurf_snow/autoassess_landsurf_snow/MPI-ESM-MR_vs_MPI-ESM-LR/land_surface_snow/MPI-ESM-MR/metrics.csv --file-ref /scratch/b/b309141/work_v2/recipe_autoassess_landsurface_snow_20180829_150609/plots/aa_landsurf_snow/autoassess_landsurf_snow/MPI-ESM-MR_vs_MPI-ESM-LR/land_surface_snow/MPI-ESM-LR/metrics.csv
/usr/bin/env: python2.7: No such file or directory
2018-08-29 17:06:39,873 [21092] INFO     __main__,61    Diagnostic (metrics plotter) has failed!
```

Can you change the code so that the diagnostic is (python) version independent? (I only have a `python` executable :smile:)

valeriupredoi commented:
great, cheers for testing @schlunma :beer: damned Met Office inheritance that py27, I will change it now, completely forgot about it since I still have py27

valeriupredoi commented:
done @schlunma -- also, you and @bouweandela will have noticed duplication reported by circleCI and she's right, I need to standardize the autoassess area diagnostic so that one will rule them all, but not now, that will be done down the line after this PR :hammer:

schlunma commented:
Hi @valeriupredoi, I tested both recipes successfully :tada: Two things I noticed:

- I got `png` files even though I use `pdf` in my configuration file.
- The `csv` and `nc` files are written to the `plots` subdirectory, I think they belong in the `work_dir`.

valeriupredoi commented:
Sweet, thank you, sir! good points, I will overhaul the whole autoassess
procedure very soon (including creating a single diag wrapper to run all
areas, keep the user's preferred output format for plots etc), right now it
still uses a bunch of legacy functionalities, hence the duplication and a
few other codacy complaints. I'd like to do this as soon as you guys ok
this PR :beer:

On Thu, 30 Aug 2018, 08:34 Manuel Schlund, <notifications@github.com> wrote:

> Hi @valeriupredoi <https://github.com/valeriupredoi>, I tested both
> recipes successfully 🎉 Two things I noticed:
>
>    - I got png files even though I use pdf in my configuration file.
>    - The csv and nc files are written to the plots subdirectory, I think
>    they belong in the work_dir.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/591#issuecomment-417219736>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AbpCo14AmaoU2yp66saidNnETRHQYKVNks5uV5WCgaJpZM4WP7A7>
> .
>

valeriupredoi commented:
you fellas are OK with this PR, I'd like to merge and immediately branch out since I want to overhaul the autoassess-area framework - no pressure :rofl:

schlunma commented:
I think @mattiarighi wants to merge new stuff himself as this requires more testing! I'm sorry V.! 😞 What's the problem with continuing your work on this branch? Also I'm not really happy about merging stuff with so many Codacy issues 😬

valeriupredoi commented:
sure no problemo, I am happy to work on this branch, only problem is that what I will get done will indeed kill off the codacy issues but it will bring in quite a bit of changes and @bouweandela and me have agreed that PR's should be small enough so reviewers don't shoot themselves :grin: This one is really not affecting anything from the core functionality, just adds diags

schlunma commented:
What's your opinion about that @bouweandela @bjoernbroetz?

bouweandela commented:
I agree with @schlunma that we should not merge PRs with many issues, because fixing things later can easily become fixing things never because it turns out you don't have the time.

@valeriupredoi To keep the PRs small, maybe you can start by refactoring the already existing code, in small steps, and then add a new feature once that is done? I also noticed that this branch contains relatively large binary files, if we merge that, the esmvaltool installation would at least double in size. Are they really necessary?

valeriupredoi commented:
sounds like a plan @bouweandela will get cracking on it on Monday, almost beer o'clock here. Thos binary files are climatology files used by the diags, for no better paths to put them I shovelled them in ESMValTool, they are problematic in the sense they should be somewhere under some sort of version control; I need to talk to the people that use them and maintain them at the Met Office see where best to put them. I do accept suggestions though, unless you say 'get rid of them' :grin:

valeriupredoi commented:
hey guys @bouweandela @schlunma this whole show has now been restructured nicely, and I wrote a unique diag wrapper for all the autoassess areas and got rid of a whole bunch of obsolete files :grin:

valeriupredoi commented:
hey guys @bouweandela @schlunma @mattiarighi - some serious updates here: I have restructured and simplified the autoassess diagnostics handling. Main highlights:

 - it is now possible to run any autoassess-native area diagnostic using the ```autoassess_area_base``` wrapper; this is not calling any intermediary scripts anymore and it just goes straight and executes the diagnostic; also, no more use of ```subprocess``` and its potential security issues; the area diag is executed simpli by importing the relevant module;
 - ```autoassess_area_base``` can be used to run any sort of diagnostic that uses a CONTROL, EXPERIMENT and OBS (obs, obs4mips etc) set of datasets, so this can actually be used by a user to run their scratchpad diagnostic that DOES NOT have to be checked in the git repo - this allows advanced users (ping @ledm ) to use ESMValTool in the first phase without having to create dev branches;
 - removed a LOT of the legacy autoassess codes;
 - streamlined the Met Office standard plot the same way as above by removing ```subprocess``` calls and executing the private ```_plot_mo_metrics``` (which is about 80% of the legacy autoassess plotter but I really don't want to change anymore of it because the Met Office is quite strict with their type of metric plot);
 - all recipes work fine - NOTE that snow uses ```GlobSnow``` climatology binary files. These files need to live in a standard path that is easily accessible by all users (just like ESGF) and are not included anymore in ESMValTool - as @bouweandela pointed out, they are big, and are enlarging the code base by a freaking lot, so for testing purposes I am attaching them here; the snow recipe includes an option that sets the rootpath to these files;
 - the codacy issues are minor now and really, I can do diddly-shit about them,

So please have a look and let me know what you think, I'd be very happy to merge this and have it behind me, plus it also will help me a great deal to include all the other autoassess area diags that still need to be plugged in, now that this is nicely standardized. Cheerios :grin: :beer: 
[GlobSnow.tar.gz](https://github.com/ESMValGroup/ESMValTool/files/2357110/GlobSnow.tar.gz)

valeriupredoi commented:
hey guys, I am very much ready to merge this, can you plss tell me if there's anything you be against :grin: @bouweandela @mattiarighi  :eggplant:

valeriupredoi commented:
hey guys @bouweandela @mattiarighi I seriously need to merge this into main, so I can start pesking the MO guys wrt missing variables in a few diags in ```REFACTORING_autoassess``` (that is now up-to-date with this branch) and would like to produce output for the CRESCENDO conference in October - I am also sick to the brim to add changes here :grinning: -- all the recipes work fine and apart from ```_supermeans``` added in ```/diag_scripts/shared``` (which is a very useful script nonetheless), all else is just in the ```autoassess``` directory. Can yous plsss have a look and OK it :beer:

mattiarighi commented:
Since this is not changing anything in the preprocessor it's fine for me to merge if @bouweandela has no further remarks.

valeriupredoi commented:
cool, cheers @mattiarighi -> @bouweandela pliz be gentle :grin:

bouweandela commented on esmvaltool/diag_scripts/autoassess/_plot_mo_metrics.py:
```diff
@@ -111,22 +87,24 @@ def write_order_metrics(csvfile, metrics):
             outf = open(csvfile, 'w')
-        except IOError as e:
-            if e.errno == errno.EACCES:
+        except IOError as shit:
```
Could you please use more professional names (also in other places)? We would not want anyone to feel offended when the read the esmvaltool source code.

valeriupredoi replied:
cheers @bouweandela this is a leftover from testing, sure thing will replace it :grinning:

valeriupredoi commented:
@bouweandela cheers for looking man, I have merged and I will fix the duplications at a later stage - they all stem from autoassess codes (not my codes) and whreas this is not a valid explanation for the duplication (I should have taken care of it) - these codes are likely to change in the near future (by me in the REFACTORING_autoassess branch)
